pkg/common/scrubber.go

```diff
@@ -0,0 +1,341 @@
+func (s *AbstractScrubber) GenerateFunctionSecretName(functionName string) string {
+	secretName := fmt.Sprintf("%s-%s", "nuclio", functionName)
+	if len(secretName) > KubernetesDomainLevelMaxLength-8 {
+		secretName = secretName[:KubernetesDomainLevelMaxLength-8]
+	}
+
+	// remove trailing non-alphanumeric characters
+	secretName = strings.TrimRight(secretName, "-_")
+
+	// add a unique id to the end of the name
+	secretName = fmt.Sprintf("%s-%s", secretName, GenerateRandomString(8, SmallLettersAndNumbers))
+
+	return secretName
+}
+
+// GenerateFlexVolumeSecretName generates a secret name for a flex volume, in the form of:
+// `nuclio-flex-volume-<volume-name>-<unique-id>`
+func (s *AbstractScrubber) GenerateFlexVolumeSecretName(functionName, volumeName string) string {
+	secretName := fmt.Sprintf("%s-%s-%s", NuclioFlexVolumeSecretNamePrefix, functionName, volumeName)
+
+	// if the secret name is too long, drop the function and project name
+	if len(secretName) > KubernetesDomainLevelMaxLength {
+		secretName = fmt.Sprintf("%s-%s", NuclioFlexVolumeSecretNamePrefix, volumeName)
+
+	}
+
+	// if the secret name is still too long, trim it and keep space for the unique id
+	if len(secretName) > KubernetesDomainLevelMaxLength-8 {
+		secretName = secretName[:KubernetesDomainLevelMaxLength-8]
+	}
+
+	// remove trailing non-alphanumeric characters
+	secretName = strings.TrimRight(secretName, "-_")
+
+	// add a unique id to the end of the name
+	secretName = fmt.Sprintf("%s-%s", secretName, GenerateRandomString(8, SmallLettersAndNumbers))
+
+	return secretName
+}
+
+// EncodeSecretKey encodes a secret key
+func (s *AbstractScrubber) EncodeSecretKey(fieldPath string) string {
+	encodedFieldPath := base64.StdEncoding.EncodeToString([]byte(fieldPath))
+	encodedFieldPath = strings.ReplaceAll(encodedFieldPath, "=", "_")
+	return fmt.Sprintf("%s%s", ReferenceToEnvVarPrefix, encodedFieldPath)
+}
+
+// DecodeSecretKey decodes a secret key and returns the original field
+func (s *AbstractScrubber) DecodeSecretKey(secretKey string) (string, error) {
+	encodedFieldPath := strings.TrimPrefix(secretKey, ReferenceToEnvVarPrefix)
+	encodedFieldPath = strings.ReplaceAll(encodedFieldPath, "_", "=")
+	decodedFieldPath, err := base64.StdEncoding.DecodeString(encodedFieldPath)
+	if err != nil {
+		return "", errors.Wrap(err, "Failed to decode secret key")
+	}
+	return string(decodedFieldPath), nil
+}
+
+func (s *AbstractScrubber) generateSecretKey(fieldPath string) string {
+	return fmt.Sprintf("%s%s", ReferencePrefix, strings.ToLower(fieldPath))
+}
+
+func (s *AbstractScrubber) ConvertMapToConfig(mapConfig interface{}) (interface{}, error) {
+	return s.Scrubber.ConvertMapToConfig(mapConfig)
+}
+
+func (s *AbstractScrubber) ValidateReference(objectToScrub interface{},
+	existingSecretMap map[string]string,
+	fieldPath,
+	secretKey,
+	stringValue string) error {
+	return s.Scrubber.ValidateReference(objectToScrub, existingSecretMap, fieldPath, secretKey, stringValue)
+}
```

Comment:
Are these really necessary in the abstract if they are implemented in the specific scrubber instances?